superset/migrations/versions/070c043f2fdb_add_granularity_to_charts_where_missing.py

```diff
@@ -0,0 +1,125 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""add granularity to charts where missing
+
+Revision ID: 070c043f2fdb
+Revises: e11ccdd12658
+Create Date: 2021-02-04 09:34:13.608891
+
+"""
+
+# revision identifiers, used by Alembic.
+revision = "070c043f2fdb"
+down_revision = "e11ccdd12658"
+
+import json
+
+from alembic import op
+from sqlalchemy import Boolean, Column, Integer, String, Text
+from sqlalchemy.ext.declarative import declarative_base
+
+from superset import db
+
+Base = declarative_base()
+
+
+class Slice(Base):
+    __tablename__ = "slices"
+
+    id = Column(Integer, primary_key=True)
+    params = Column(Text)
+    datasource_id = Column(Integer)
+    datasource_type = Column(String(200))
+
+
+class SqlaTable(Base):
+    __tablename__ = "tables"
+
+    id = Column(Integer, primary_key=True)
+    main_dttm_col = Column(String(250))
+
+
+class TableColumn(Base):
+    __tablename__ = "table_columns"
+
+    id = Column(Integer, primary_key=True)
+    table_id = Column(Integer)
+    is_dttm = Column(Boolean)
+    column_name = Column(String(255))
+
+
+def upgrade():
+    """
+    Adds the granularity param to charts without it populated. This is required for
+    time range filtering to work properly. Uses the following approach:
+
+    - Find all charts without a granularity or granularity_sqla param.
+    - Get the dataset that backs the chart.
+    - If the dataset has the main dttm column set, use it.
+    - Otherwise, find all the dttm columns in the dataset and use the first one.
+    - If no dttm columns exist in the dataset, don't change the chart.
+    """
+    bind = op.get_bind()
+    session = db.Session(bind=bind)
+
+    slices_changed = 0
+
+    for slc in session.query(Slice).filter(Slice.datasource_type == "table").all():
+        try:
+            params = json.loads(slc.params)
+            if "granularity" in params or "granularity_sqla" in params:
+                continue
+
+            table = session.query(SqlaTable).get(slc.datasource_id)
```

Comment:
Is this performant? I wonder if the join should be part of the slice query.

Reply:
This ended up only altering 150 slices in our DB (and only got to this step for about 4k) so i'm not sure performance matters that much. It's a trade off between doing the join with a much larger number of slices (200k+) vs. waiting until we get to this step. idk